File: sdk/python/trueflow/resources/secret_references.py
# Secret References resource

"""Resource for managing external secret references."""

from typing import List, Optional, Dict, Any
from datetime import datetime

from pydantic import BaseModel, Field

from ..exceptions import raise_for_status


# ============================================================================
# Pydantic Models
# ============================================================================

class SecretReference(BaseModel):
    """A reference to an external secret in a vault backend.

    Secret references provide a workspace-scoped abstraction over external
    secrets stored in AWS Secrets Manager, HashiCorp Vault KV, or Azure Key Vault.
    """
    id: str
    project_id: str
    name: str
    description: Optional[str] = None
    vault_backend: str
    external_ref: str
    vault_config_id: Optional[str] = None
    provider: Optional[str] = None
    injection_mode: str
    injection_header: str
    allowed_team_ids: Optional[List[str]] = None
    allowed_user_ids: Optional[List[str]] = None
    last_accessed_at: Optional[datetime] = None
    last_rotated_at: Optional[datetime] = None
    version: Optional[str] = None
    is_active: bool
    created_at: datetime
    updated_at: datetime
    created_by: Optional[str] = None

    def __repr__(self) -> str:
        return f"SecretReference(id={self.id!r}, name={self.name!r}, vault_backend={self.vault_backend!r})"


class SecretReferenceCreateResponse(BaseModel):
    """Response from creating a secret reference."""
    id: Optional[str] = None
    name: Optional[str] = None
    vault_backend: Optional[str] = None
    external_ref: Optional[str] = None


class SecretFetchResponse(BaseModel):
    """Response from fetching a secret."""
    reference_id: str
    fetched_at: datetime
    message: str


# ============================================================================
# Sync Resource
# ============================================================================

class SecretReferencesResource:
    """Management API resource for secret references."""

    def __init__(self, client):
        self._client = client

    def list(
        self,
        *,
        vault_backend: Optional[str] = None,
        provider: Optional[str] = None,
        is_active: Optional[bool] = None,
        limit: Optional[int] = None,
        offset: Optional[int] = None,
    ) -> List[SecretReference]:
        """List secret references.

        Args:
            vault_backend: Filter by vault backend type
                (aws_secrets_manager, hashicorp_vault, hashicorp_vault_kv, azure_key_vault).
            provider: Filter by provider (e.g., "openai", "anthropic").
            is_active: Filter by active status.
            limit: Maximum number of results (default 100, max 1000).
            offset: Pagination offset.

        Returns:
            List of SecretReference objects.
        """
        params: Dict[str, Any] = {}
        if vault_backend:
            params["vault_backend"] = vault_backend
        if provider:
            params["provider"] = provider
        if is_active is not None:
            params["is_active"] = is_active
        if limit is not None:
            params["limit"] = limit
        if offset is not None:
            params["offset"] = offset

        resp = self._client._http.get("/api/v1/secret-references", params=params)
        raise_for_status(resp)
        return [SecretReference(**item) for item in resp.json()]

    def create(
        self,
        name: str,
        vault_backend: str,
        external_ref: str,
        *,
        description: Optional[str] = None,
        vault_config_id: Optional[str] = None,
        provider: Optional[str] = None,
        injection_mode: Optional[str] = None,
        injection_header: Optional[str] = None,
        allowed_team_ids: Optional[List[str]] = None,
        allowed_user_ids: Optional[List[str]] = None,
    ) -> SecretReference:
        """Create a new secret reference.

        Args:
            name: Human-readable name for the secret reference.
            vault_backend: Vault backend type. One of: aws_secrets_manager,
                hashicorp_vault, hashicorp_vault_kv, azure_key_vault.
            external_ref: External reference to the secret.
                For AWS: ARN (e.g., "arn:aws:secretsmanager:us-east-1:123456789012:secret:my-secret")
                For HashiCorp Vault: path:key format (e.g., "secret/data/api:openai_key")
                For Azure Key Vault: secret URI.
            description: Optional description.
            vault_config_id: Optional vault config ID for authentication.
            provider: Provider this secret is for (e.g., "openai", "anthropic").
            injection_mode: How to inject the secret. One of: bearer (default),
                header, query, none.
            injection_header: Header name for injection (default: Authorization).
            allowed_team_ids: Team IDs allowed to access this secret.
            allowed_user_ids: User IDs allowed to access this secret.

        Returns:
            The created SecretReference.

        Example:
            ref = client.secret_references.create(
                name="prod-openai-key",
                vault_backend="aws_secrets_manager",
                external_ref="arn:aws:secretsmanager:us-east-1:123456789012:secret:prod/openai-key",
                provider="openai",
            )
        """
        payload: Dict[str, Any] = {
            "name": name,
            "vault_backend": vault_backend,
            "external_ref": external_ref,
        }
        if description is not None:
            payload["description"] = description
        if vault_config_id is not None:
            payload["vault_config_id"] = vault_config_id
        if provider is not None:
            payload["provider"] = provider
        if injection_mode is not None:
            payload["injection_mode"] = injection_mode
        if injection_header is not None:
            payload["injection_header"] = injection_header
        if allowed_team_ids is not None:
            payload["allowed_team_ids"] = allowed_team_ids
        if allowed_user_ids is not None:
            payload["allowed_user_ids"] = allowed_user_ids

        resp = self._client._http.post("/api/v1/secret-references", json=payload)
        raise_for_status(resp)
        return SecretReference(**resp.json())

    def get(self, id: str) -> SecretReference:
        """Get a secret reference by ID.

        Args:
            id: The secret reference ID.

        Returns:
            The SecretReference object.
        """
        resp = self._client._http.get(f"/api/v1/secret-references/{id}")
        raise_for_status(resp)
        return SecretReference(**resp.json())

    def update(
        self,
        id: str,
        *,
        name: Optional[str] = None,
        description: Optional[str] = None,
        external_ref: Optional[str] = None,
        vault_config_id: Optional[str] = None,
        provider: Optional[str] = None,
        injection_mode: Optional[str] = None,
        injection_header: Optional[str] = None,
        allowed_team_ids: Optional[List[str]] = None,
        allowed_user_ids: Optional[List[str]] = None,
        version: Optional[str] = None,
        is_active: Optional[bool] = None,
    ) -> SecretReference:
        """Update a secret reference.

        Args:
            id: The secret reference ID.
            name: New name for the secret reference.
            description: New description.
            external_ref: New external reference.
            vault_config_id: New vault config ID.
            provider: New provider.
            injection_mode: New injection mode.
            injection_header: New injection header.
            allowed_team_ids: New list of allowed team IDs.
            allowed_user_ids: New list of allowed user IDs.
            version: Secret version (for vaults that support versioning).
            is_active: Active status.

        Returns:
            The updated SecretReference.
        """
        payload: Dict[str, Any] = {}
        if name is not None:
            payload["name"] = name
        if description is not None:
            payload["description"] = description
        if external_ref is not None:
            payload["external_ref"] = external_ref
        if vault_config_id is not None:
            payload["vault_config_id"] = vault_config_id
        if provider is not None:
            payload["provider"] = provider
        if injection_mode is not None:
            payload["injection_mode"] = injection_mode
        if injection_header is not None:
            payload["injection_header"] = injection_header
        if allowed_team_ids is not None:
            payload["allowed_team_ids"] = allowed_team_ids
        if allowed_user_ids is not None:
            payload["allowed_user_ids"] = allowed_user_ids
        if version is not None:
            payload["version"] = version
        if is_active is not None:
            payload["is_active"] = is_active

        resp = self._client._http.put(f"/api/v1/secret-references/{id}", json=payload)
        raise_for_status(resp)
        return SecretReference(**resp.json())

    def delete(self, id: str) -> Dict[str, Any]:
        """Delete a secret reference.

        Args:
            id: The secret reference ID.

        Returns:
            Dict with 'id' and 'deleted' fields.
        """
        resp = self._client._http.delete(f"/api/v1/secret-references/{id}")
        raise_for_status(resp)
        return resp.json()

    def fetch(self, id: str) -> SecretFetchResponse:
        """Fetch and cache the secret from the external vault.

        This triggers an immediate fetch of the secret from the external vault.
        The secret is cached in Redis for subsequent lookups. The actual secret
        value is not returned in the response.

        Args:
            id: The secret reference ID.

        Returns:
            SecretFetchResponse with fetch status.
        """
        resp = self._client._http.post(f"/api/v1/secret-references/{id}/fetch")
        raise_for_status(resp)
        return SecretFetchResponse(**resp.json())


# ============================================================================
# Async Resource
# ============================================================================

class AsyncSecretReferencesResource:
    """Async Management API resource for secret references."""

    def __init__(self, client):
        self._client = client

    async def list(
        self,
        *,
        vault_backend: Optional[str] = None,
        provider: Optional[str] = None,
        is_active: Optional[bool] = None,
        limit: Optional[int] = None,
        offset: Optional[int] = None,
    ) -> List[SecretReference]:
        """List secret references.

        Args:
            vault_backend: Filter by vault backend type
                (aws_secrets_manager, hashicorp_vault, hashicorp_vault_kv, azure_key_vault).
            provider: Filter by provider (e.g., "openai", "anthropic").
            is_active: Filter by active status.
            limit: Maximum number of results (default 100, max 1000).
            offset: Pagination offset.

        Returns:
            List of SecretReference objects.
        """
        params: Dict[str, Any] = {}
        if vault_backend:
            params["vault_backend"] = vault_backend
        if provider:
            params["provider"] = provider
        if is_active is not None:
            params["is_active"] = is_active
        if limit is not None:
            params["limit"] = limit
        if offset is not None:
            params["offset"] = offset

        resp = await self._client._http.get("/api/v1/secret-references", params=params)
        raise_for_status(resp)
        return [SecretReference(**item) for item in resp.json()]

    async def create(
        self,
        name: str,
        vault_backend: str,
        external_ref: str,
        *,
        description: Optional[str] = None,
        vault_config_id: Optional[str] = None,
        provider: Optional[str] = None,
        injection_mode: Optional[str] = None,
        injection_header: Optional[str] = None,
        allowed_team_ids: Optional[List[str]] = None,
        allowed_user_ids: Optional[List[str]] = None,
    ) -> SecretReference:
        """Create a new secret reference.

        Args:
            name: Human-readable name for the secret reference.
            vault_backend: Vault backend type. One of: aws_secrets_manager,
                hashicorp_vault, hashicorp_vault_kv, azure_key_vault.
            external_ref: External reference to the secret.
                For AWS: ARN (e.g., "arn:aws:secretsmanager:us-east-1:123456789012:secret:my-secret")
                For HashiCorp Vault: path:key format (e.g., "secret/data/api:openai_key")
                For Azure Key Vault: secret URI.
            description: Optional description.
            vault_config_id: Optional vault config ID for authentication.
            provider: Provider this secret is for (e.g., "openai", "anthropic").
            injection_mode: How to inject the secret. One of: bearer (default),
                header, query, none.
            injection_header: Header name for injection (default: Authorization).
            allowed_team_ids: Team IDs allowed to access this secret.
            allowed_user_ids: User IDs allowed to access this secret.

        Returns:
            The created SecretReference.
        """
        payload: Dict[str, Any] = {
            "name": name,
            "vault_backend": vault_backend,
            "external_ref": external_ref,
        }
        if description is not None:
            payload["description"] = description
        if vault_config_id is not None:
            payload["vault_config_id"] = vault_config_id
        if provider is not None:
            payload["provider"] = provider
        if injection_mode is not None:
            payload["injection_mode"] = injection_mode
        if injection_header is not None:
            payload["injection_header"] = injection_header
        if allowed_team_ids is not None:
            payload["allowed_team_ids"] = allowed_team_ids
        if allowed_user_ids is not None:
            payload["allowed_user_ids"] = allowed_user_ids

        resp = await self._client._http.post("/api/v1/secret-references", json=payload)
        raise_for_status(resp)
        return SecretReference(**resp.json())

    async def get(self, id: str) -> SecretReference:
        """Get a secret reference by ID.

        Args:
            id: The secret reference ID.

        Returns:
            The SecretReference object.
        """
        resp = await self._client._http.get(f"/api/v1/secret-references/{id}")
        raise_for_status(resp)
        return SecretReference(**resp.json())

    async def update(
        self,
        id: str,
        *,
        name: Optional[str] = None,
        description: Optional[str] = None,
        external_ref: Optional[str] = None,
        vault_config_id: Optional[str] = None,
        provider: Optional[str] = None,
        injection_mode: Optional[str] = None,
        injection_header: Optional[str] = None,
        allowed_team_ids: Optional[List[str]] = None,
        allowed_user_ids: Optional[List[str]] = None,
        version: Optional[str] = None,
        is_active: Optional[bool] = None,
    ) -> SecretReference:
        """Update a secret reference.

        Args:
            id: The secret reference ID.
            name: New name for the secret reference.
            description: New description.
            external_ref: New external reference.
            vault_config_id: New vault config ID.
            provider: New provider.
            injection_mode: New injection mode.
            injection_header: New injection header.
            allowed_team_ids: New list of allowed team IDs.
            allowed_user_ids: New list of allowed user IDs.
            version: Secret version (for vaults that support versioning).
            is_active: Active status.

        Returns:
            The updated SecretReference.
        """
        payload: Dict[str, Any] = {}
        if name is not None:
            payload["name"] = name
        if description is not None:
            payload["description"] = description
        if external_ref is not None:
            payload["external_ref"] = external_ref
        if vault_config_id is not None:
            payload["vault_config_id"] = vault_config_id
        if provider is not None:
            payload["provider"] = provider
        if injection_mode is not None:
            payload["injection_mode"] = injection_mode
        if injection_header is not None:
            payload["injection_header"] = injection_header
        if allowed_team_ids is not None:
            payload["allowed_team_ids"] = allowed_team_ids
        if allowed_user_ids is not None:
            payload["allowed_user_ids"] = allowed_user_ids
        if version is not None:
            payload["version"] = version
        if is_active is not None:
            payload["is_active"] = is_active

        resp = await self._client._http.put(f"/api/v1/secret-references/{id}", json=payload)
        raise_for_status(resp)
        return SecretReference(**resp.json())

    async def delete(self, id: str) -> Dict[str, Any]:
        """Delete a secret reference.

        Args:
            id: The secret reference ID.

        Returns:
            Dict with 'id' and 'deleted' fields.
        """
        resp = await self._client._http.delete(f"/api/v1/secret-references/{id}")
        raise_for_status(resp)
        return resp.json()

    async def fetch(self, id: str) -> SecretFetchResponse:
        """Fetch and cache the secret from the external vault.

        This triggers an immediate fetch of the secret from the external vault.
        The secret is cached in Redis for subsequent lookups. The actual secret
        value is not returned in the response.

        Args:
            id: The secret reference ID.

        Returns:
            SecretFetchResponse with fetch status.
        """
        resp = await self._client._http.post(f"/api/v1/secret-references/{id}/fetch")
        raise_for_status(resp)
        return SecretFetchResponse(**resp.json())

File: gateway/src/api/handlers/secret_references.rs
//! Secret Reference API handlers.
//!
//! These endpoints manage workspace-scoped references to external secrets
//! stored in AWS Secrets Manager, HashiCorp Vault KV, or Azure Key Vault.

use std::sync::Arc;

use axum::{
    extract::{Path, Query, State},
    http::StatusCode,
    Extension, Json,
};
use uuid::Uuid;

use super::dtos::{
    CreateSecretReferenceDto, DeleteResponse, SecretFetchResponse, SecretReferenceFilterParams,
    SecretReferenceResponse, UpdateSecretReferenceDto,
};
use super::helpers::verify_project_ownership;
use crate::api::AuthContext;
use crate::models::secret_reference::{
    CreateSecretReferenceRequest, SecretReferenceFilter, UpdateSecretReferenceRequest,
};
use crate::AppState;

/// GET /api/v1/secret-references — list secret references
pub async fn list_secret_references(
    State(state): State<Arc<AppState>>,
    Extension(auth): Extension<AuthContext>,
    Query(params): Query<SecretReferenceFilterParams>,
) -> Result<Json<Vec<SecretReferenceResponse>>, StatusCode> {
    auth.require_scope("credentials:read")
        .map_err(|_| StatusCode::FORBIDDEN)?;

    let project_id = auth.default_project_id();
    verify_project_ownership(&state, auth.org_id, project_id).await?;

    let filter = SecretReferenceFilter {
        vault_backend: params.vault_backend,
        provider: params.provider,
        is_active: params.is_active,
    };

    let limit = params.limit.unwrap_or(100).min(1000);
    let offset = params.offset.unwrap_or(0);

    let refs = state
        .db
        .list_secret_references(project_id, filter, limit, offset)
        .await
        .map_err(|e| {
            tracing::error!("list_secret_references failed: {}", e);
            StatusCode::INTERNAL_SERVER_ERROR
        })?;

    Ok(Json(refs.into_iter().map(Into::into).collect()))
}

/// POST /api/v1/secret-references — create a new secret reference
pub async fn create_secret_reference(
    State(state): State<Arc<AppState>>,
    Extension(auth): Extension<AuthContext>,
    Json(payload): Json<CreateSecretReferenceDto>,
) -> Result<(StatusCode, Json<SecretReferenceResponse>), StatusCode> {
    auth.require_role("admin")?;
    auth.require_scope("credentials:write")
        .map_err(|_| StatusCode::FORBIDDEN)?;

    let project_id = auth.default_project_id();
    verify_project_ownership(&state, auth.org_id, project_id).await?;

    // Validate injection mode if provided
    if let Some(ref mode) = payload.injection_mode {
        match mode.as_str() {
            "bearer" | "header" | "query" | "none" => {}
            _ => {
                tracing::warn!("create_secret_reference: invalid injection_mode: {}", mode);
                return Err(StatusCode::BAD_REQUEST);
            }
        }
    }

    // Validate injection header name if provided
    if let Some(ref header) = payload.injection_header {
        if reqwest::header::HeaderName::from_bytes(header.as_bytes()).is_err() {
            tracing::warn!("create_secret_reference: invalid injection_header: {}", header);
            return Err(StatusCode::BAD_REQUEST);
        }
    }

    // Validate vault_backend for external vaults only
    match payload.vault_backend.as_str() {
        "aws_secrets_manager" | "hashicorp_vault" | "hashicorp_vault_kv" | "azure_key_vault" => {}
        _ => {
            tracing::warn!("create_secret_reference: invalid vault_backend: {}", payload.vault_backend);
            return Err(StatusCode::BAD_REQUEST);
        }
    }

    let req = CreateSecretReferenceRequest {
        name: payload.name,
        description: payload.description,
        vault_backend: payload.vault_backend,
        external_ref: payload.external_ref,
        vault_config_id: payload.vault_config_id,
        provider: payload.provider,
        injection_mode: payload.injection_mode,
        injection_header: payload.injection_header,
        allowed_team_ids: payload.allowed_team_ids,
        allowed_user_ids: payload.allowed_user_ids,
    };

    let sr = state
        .db
        .create_secret_reference(project_id, req, auth.user_id)
        .await
        .map_err(|e| {
            tracing::error!("create_secret_reference failed: {}", e);
            StatusCode::INTERNAL_SERVER_ERROR
        })?;

    Ok((StatusCode::CREATED, Json(sr.into())))
}

/// GET /api/v1/secret-references/:id — get a secret reference
pub async fn get_secret_reference(
    State(state): State<Arc<AppState>>,
    Extension(auth): Extension<AuthContext>,
    Path(id_str): Path<String>,
) -> Result<Json<SecretReferenceResponse>, StatusCode> {
    auth.require_scope("credentials:read")
        .map_err(|_| StatusCode::FORBIDDEN)?;

    let id = Uuid::parse_str(&id_str).map_err(|_| StatusCode::BAD_REQUEST)?;
    let project_id = auth.default_project_id();
    verify_project_ownership(&state, auth.org_id, project_id).await?;

    let sr = state
        .db
        .get_secret_reference(id, project_id)
        .await
        .map_err(|e| {
            tracing::error!("get_secret_reference failed: {}", e);
            StatusCode::INTERNAL_SERVER_ERROR
        })?
        .ok_or(StatusCode::NOT_FOUND)?;

    Ok(Json(sr.into()))
}

/// PUT /api/v1/secret-references/:id — update a secret reference
pub async fn update_secret_reference(
    State(state): State<Arc<AppState>>,
    Extension(auth): Extension<AuthContext>,
    Path(id_str): Path<String>,
    Json(payload): Json<UpdateSecretReferenceDto>,
) -> Result<Json<SecretReferenceResponse>, StatusCode> {
    auth.require_role("admin")?;
    auth.require_scope("credentials:write")
        .map_err(|_| StatusCode::FORBIDDEN)?;

    let id = Uuid::parse_str(&id_str).map_err(|_| StatusCode::BAD_REQUEST)?;
    let project_id = auth.default_project_id();
    verify_project_ownership(&state, auth.org_id, project_id).await?;

    // Validate injection mode if provided
    if let Some(ref mode) = payload.injection_mode {
        match mode.as_str() {
            "bearer" | "header" | "query" | "none" => {}
            _ => {
                tracing::warn!("update_secret_reference: invalid injection_mode: {}", mode);
                return Err(StatusCode::BAD_REQUEST);
            }
        }
    }

    // Validate injection header name if provided
    if let Some(ref header) = payload.injection_header {
        if reqwest::header::HeaderName::from_bytes(header.as_bytes()).is_err() {
            tracing::warn!("update_secret_reference: invalid injection_header: {}", header);
            return Err(StatusCode::BAD_REQUEST);
        }
    }

    let req = UpdateSecretReferenceRequest {
        name: payload.name,
        description: payload.description,
        external_ref: payload.external_ref,
        vault_config_id: payload.vault_config_id,
        provider: payload.provider,
        injection_mode: payload.injection_mode,
        injection_header: payload.injection_header,
        allowed_team_ids: payload.allowed_team_ids,
        allowed_user_ids: payload.allowed_user_ids,
        version: payload.version,
        is_active: payload.is_active,
    };

    let sr = state
        .db
        .update_secret_reference(id, project_id, req)
        .await
        .map_err(|e| {
            tracing::error!("update_secret_reference failed: {}", e);
            StatusCode::INTERNAL_SERVER_ERROR
        })?
        .ok_or(StatusCode::NOT_FOUND)?;

    Ok(Json(sr.into()))
}

/// DELETE /api/v1/secret-references/:id — delete a secret reference
pub async fn delete_secret_reference(
    State(state): State<Arc<AppState>>,
    Extension(auth): Extension<AuthContext>,
    Path(id_str): Path<String>,
) -> Result<Json<DeleteResponse>, StatusCode> {
    auth.require_role("admin")?;
    auth.require_scope("credentials:write")
        .map_err(|_| StatusCode::FORBIDDEN)?;

    let id = Uuid::parse_str(&id_str).map_err(|_| StatusCode::BAD_REQUEST)?;
    let project_id = auth.default_project_id();
    verify_project_ownership(&state, auth.org_id, project_id).await?;

    let deleted = state
        .db
        .hard_delete_secret_reference(id, project_id)
        .await
        .map_err(|e| {
            tracing::error!("delete_secret_reference failed: {}", e);
            StatusCode::INTERNAL_SERVER_ERROR
        })?;

    if !deleted {
        return Err(StatusCode::NOT_FOUND);
    }

    Ok(Json(DeleteResponse { id, deleted }))
}

/// POST /api/v1/secret-references/:id/fetch — fetch and cache the secret value
///
/// This endpoint triggers an immediate fetch of the secret from the external vault.
/// The secret is cached in Redis for subsequent lookups.
pub async fn fetch_secret(
    State(state): State<Arc<AppState>>,
    Extension(auth): Extension<AuthContext>,
    Path(id_str): Path<String>,
) -> Result<Json<SecretFetchResponse>, StatusCode> {
    auth.require_scope("credentials:read")
        .map_err(|_| StatusCode::FORBIDDEN)?;

    let id = Uuid::parse_str(&id_str).map_err(|_| StatusCode::BAD_REQUEST)?;
    let project_id = auth.default_project_id();
    verify_project_ownership(&state, auth.org_id, project_id).await?;

    // Get the secret reference
    let sr = state
        .db
        .get_secret_reference(id, project_id)
        .await
        .map_err(|e| {
            tracing::error!("fetch_secret: get_secret_reference failed: {}", e);
            StatusCode::INTERNAL_SERVER_ERROR
        })?
        .ok_or(StatusCode::NOT_FOUND)?;

    // Check if the secret reference is active
    if !sr.is_active {
        return Err(StatusCode::BAD_REQUEST);
    }

    // Check access if allowlists are configured
    // User must be authenticated for access control
    let user_id = auth.user_id.ok_or_else(|| {
        tracing::warn!("fetch_secret: unauthenticated access attempt");
        StatusCode::UNAUTHORIZED
    })?;

    let access_result = state
        .db
        .check_secret_access(id, user_id, None)
        .await
        .map_err(|e| {
            tracing::error!("fetch_secret: check_secret_access failed: {}", e);
            StatusCode::INTERNAL_SERVER_ERROR
        })?;

    match access_result {
        crate::models::secret_reference::SecretAccessResult::Granted => {}
        crate::models::secret_reference::SecretAccessResult::Denied => {
            tracing::warn!(
                secret_reference_id = %id,
                user_id = ?auth.user_id,
                "Secret access denied"
            );
            return Err(StatusCode::FORBIDDEN);
        }
        crate::models::secret_reference::SecretAccessResult::NotFound => {
            return Err(StatusCode::NOT_FOUND);
        }
    }

    // Fetch the secret from the vault using the cached retrieval
    // Note: The actual secret value is intentionally not returned in the response
    let (_secret, provider, injection_mode, injection_header) = state
        .vault
        .retrieve_credential_cached(state.db.pool(), &sr.external_ref)
        .await
        .map_err(|e| {
            tracing::error!("fetch_secret: vault fetch failed: {}", e);
            StatusCode::INTERNAL_SERVER_ERROR
        })?;

    // Update last_accessed_at
    if let Err(e) = state.db.touch_secret_access(id).await {
        tracing::warn!("fetch_secret: failed to update last_accessed_at: {}", e);
    }

    // Cache the secret for future lookups (optional, for performance)
    let cache_key = format!("secret_ref:{}", id);
    let cached_data = serde_json::json!({
        "provider": provider,
        "injection_mode": injection_mode,
        "injection_header": injection_header,
    });
    if let Err(e) = state.cache.set(&cache_key, &cached_data, 3600).await {
        tracing::warn!("fetch_secret: failed to cache secret metadata: {}", e);
    }

    Ok(Json(SecretFetchResponse {
        reference_id: id,
        fetched_at: chrono::Utc::now(),
        message: "Secret fetched and cached successfully".to_string(),
    }))
}